includes/traits/llms-trait-singleton.php

```diff
@@ -0,0 +1,54 @@
+<?php
+/**
+ * LifterLMS singleton trait
+ *
+ * @package LifterLMS/Traits
+ *
+ * @since [version]
+ * @version [version]
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * LifterLMS singleton trait.
+ *
+ * @since [version]
+ */
+trait LLMS_Trait_Singleton {
+
+	/**
+	 * Singleton instance of the class.
+	 *
+	 * @var object
+	 */
+	private static $instance = null;
+
+	/**
+	 * Returns a singleton instance of the class that uses this trait.
+	 *
+	 * @since [version]
+	 *
+	 * @return self
+	 */
+	public static function instance() {
+
+		if ( is_null( self::$instance ) ) {
+
+			if ( property_exists( __CLASS__, '_instance') ) {
+
+				if ( is_null( self::$_instance ) ) {
+
+					self::$instance = new self();
+					self::$_instance = self::$instance;
+				} else {
+					self::$instance = self::$_instance;
+				}
+			} else {
+				self::$instance = new self();
+			}
+		}
```

Comment:
I might be missing something but I don't think the the `is_null()` condition on line 40 is necessary. This should get the job done the same, right?

The idea being that `$_instance` should be set to be equal to `$instance` when the property exists. EG: these two var should always be either both `null` or both equal to the same instance so an additional `is_null()` on the `_instance` var shouldn't ever be necessary.

```suggestion
		if ( is_null( self::$instance ) ) {
			self::$instance = new self();
			if ( property_exists( __CLASS__, '_instance' ) ) {
				self::$_instance = self::$instance;
			}
		}
```

Reply:
If a 3rd party developer extends a LifterLMS class  with the `$_instance` property, **and** their code sets it, then we should use what is in `$_instance`. I know that this is a highly unlikely edge case.

Reply:
You're right but I hate it. :+1: